### PR TITLE
fix(injectors): sync injector tree with dom element tree.

### LINKED ISSUE #2498: ViewContainer does not hook up injectors correctly
When we have 
```
<cmp *ng-if="'true'"></cmp>
```

The injector creating component cmp is a child of the injector containing `NgIf` (in fact NgIf is injectable into cmp. The injector tree should match the DOM tree and since the DOM looks like

```
<template/>
<cmp/>
```
The cmp injector is a sibling of the ng-if injector.

### PR DESCRIPTION
**Pull request by rkirov**
Changes createFixedSize method to allow for growable lists in Dart too.

Closes: #2498

### CODE REVIEW

File: modules/angular2/src/core/compiler/view_manager_utils.ts [hunk 1]
```diff
@@ -115,7 +120,12 @@ export class AppViewManagerUtils {
     view.changeDetector.remove();
     ListWrapper.removeAt(viewContainer.views, atIndex);
     for (var i = 0; i < view.rootElementInjectors.length; ++i) {
-      view.rootElementInjectors[i].unlink();
+      var inj = view.rootElementInjectors[i];
+      if (isPresent(inj.parent)) {
+        inj.unlink();
+      } else {
+        parentView.rootElementInjectors.pop(inj);
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/view_manager_utils.ts]: This is not correct as `pop` does not take any arguments.
You should get the index in the `rootElementInjectors` and then `splice` it...

File: modules/angular2/src/facade/collection.dart [hunk 2]
```diff
@@ -92,7 +92,10 @@ class StringMapWrapper {
 
 class ListWrapper {
   static List clone(Iterable l) => new List.from(l);
-  static List createFixedSize(int size) => new List(size);
+  // new List(length) creates a non-growable list. To match JS semantics we use
+  // List.filled
+  static List createFixedSize(int size) =>
```
Comment (tbosch) [inline modules/angular2/src/facade/collection.dart]: I would rather like to have 2 methods:
- `createFixedSize` as before with actual fixed size. This is what we use most of the time and using a non growable list in Dart is probably faster.
- `createGroableSize` which is what you implemented here.

File: modules/angular2/test/core/compiler/view_manager_utils_spec.ts [hunk 3]
```diff
@@ -202,8 +202,7 @@ export function main() {
          () => {
            createViews();
            utils.attachViewInContainer(parentView, 0, contextView, 0, 0, childView);
-           expect(childView.rootElementInjectors[0].spy('linkAfter'))
-               .toHaveBeenCalledWith(contextView.elementInjectors[0], null);
+           expect(contextView.elementInjectors.length).toEqual(2);
```
Comment (tbosch) [inline modules/angular2/test/core/compiler/view_manager_utils_spec.ts]: Could you add a new test with the additional case here?

File: modules/angular2/test/core/compiler/query_integration_spec.ts [hunk 4]
```diff
@@ -76,14 +76,11 @@ export function main() {
...
 
-      // TODO(rado): The test below should be using descendants: false,
-      // but due to a bug with how injectors are hooked up query considers the
-      // directives to be distances 2 instead of direct children.
       it('should reflect dynamically inserted directives',
          inject([TestBed, AsyncTestCompleter], (tb: TestBed, async) => {
            var template =
                '<div text="1"></div>' +
-               '<needs-query-desc text="2"><div *ng-if="shouldShow" [text]="\'3\'"></div></needs-query-desc>' +
+               '<needs-query text="2"><div *ng-if="shouldShow" [text]="\'3\'"></div></needs-query>' +
```
Comment (tbosch) [inline modules/angular2/test/core/compiler/query_integration_spec.ts]: Is the default that we query on the element and one level deep elements?

**Comment by rkirov** [comment]
merged

**Pull request closed without merging**